[WIP]: Use utils::directories instead of db::config to get dirs

pwrobelse opened the pull request:
This change:
 - extends the public interface of utils::directories class to provide required directory paths to the users
 - removes 'db::config::setup_directories()' to avoid altering the fields of configuration object
 - replaces usages of db::config object with utils::directories object in places that require obtaining paths to dirs

Refs: scylladb#5626

pwrobelse commented:
This PR is a draft, because the change is in WIP state. A few things are still missing including adjustments of usage of `db::config` and `utils::directories` in tests.

It has been created to get feedback and ensure, that the direction is correct.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5810/)
#### Build Details:
- Duration: 1 hr 4 min
- Builder: i-0b378fc589fc1c6f6 (m5d.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5813/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5813/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7181/)
#### Build Details:
- Duration: 2 hr 9 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
@xemul please review

pwrobelse commented on utils/directories.cc:
```diff
@@ -145,4 +149,128 @@ future<> directories::verify_owner_and_mode(fs::path path, recursive recursive)
+}
+
+void directories::setup_paths(const db::config& cfg) {
+    if (cfg.work_directory.is_set()) {
```
I will add a test to verify that this condition does not change the logic that was present before.

`work_directory` may have a default value: `"/var/lib/scylla"`. In `db::config::setup_directories()` the predicate that checks if the variable is set, was not called for this variable.

```
bool is_set() const {
    return _source > config_source::None;
}
```

pwrobelse replied:
The test showed, that it was an issue. It has been fixed and a new test has been created.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5829/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5829/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7195/)
#### Build Details:
- Duration: 2 hr 27 min
- Builder: i-04ef12775eae1391e (m5d.12xlarge)

xemul commented on replica/database.hh:
```diff
@@ -1749,6 +1752,11 @@ public:
         return _cfg;
     }
+
+    const utils::directories& get_dirs() const {
```
database must _not_ be a provider for directories. If any code needs directory, it should get its reference on start, not borrow it via one if its dependent service

pwrobelse replied:
Sure, fixed :+1:

xemul commented on test/lib/cql_test_env.cc:
```diff
@@ -75,11 +77,6 @@
-
-
-} // anonymous namespace
-
```
Such cleanups are welcome as independent pull requests

pwrobelse replied:
Sure, reverted :+1:

xemul commented on main.cc:
```diff
@@ -966,14 +966,13 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            dir_set.add(dirs->get_data_file_dirs());
+            dir_set.add(dirs->get_commitlog_dir());
+            dir_set.add(dirs->get_schema_commitlog_dir());
             dirs->create_and_verify(std::move(dir_set)).get();
```
This thing becomes buggy now. You have already captured all the directories from cfg in the constructor up above (the `dirs.emplace(cfg)` part). Now the dir_set duplicates the set, I think that the set should be replaced in favor of just using the specific dirs from directories themselves

pwrobelse replied:
I had similar feeling about that. I was just not sure how to proceed. I preserved the logic.

1. Three field are used here to create directories via `utils::directories::set` + `create_and_verify()`.
2. Two fields are used below to create directories via `db::hints::directory_initializer` + `ensure_created_and_verified()`.

Therefore, I was not sure whether:
- `utils::directories` should create directories only for these three fields and the remaining two should be created via `directory_initializer` as it was
- `utils::directories` should create directories for all five fields and the usage of `directory_initializer` should be altered

I do not have a strong opinion, because I am not aware of many aspects related to the project. What do you think about it?

pwrobelse replied:
I checked and it seems, that one of hints dir may be created conditionally.

In `main.cc` we check `hinted_handoff_enabled.is_disabled_for_all` and depending on that flag we create the directories.

In `api/storage_proxy.cc` I can see that the flag may be modified via request:
```
    sp::set_hinted_handoff_enabled.set(r, [&proxy](std::unique_ptr<http::request> req)  {
        auto enable = req->get_query_param("enable");
        auto filter = (enable == "true" || enable == "1")
                ? db::hints::host_filter(db::hints::host_filter::enabled_for_all_tag {})
                : db::hints::host_filter(db::hints::host_filter::disabled_for_all_tag {});
        return proxy.invoke_on_all([filter = std::move(filter)] (service::storage_proxy& sp) {
            return sp.change_hints_host_filter(filter);
        }).then([] {
            return make_ready_future<json::json_return_type>(json_void());
        });
    });
```

The handler for this request calls `sp.change_hints_host_filter()`. In case of storage proxy it may create the directories via dir initializer.

```
future<> storage_proxy::change_hints_host_filter(db::hints::host_filter new_filter) {
    if (new_filter == _hints_manager.get_host_filter()) {
        co_return;
    }

    co_await _hints_directory_initializer.ensure_created_and_verified();
    co_await _hints_directory_initializer.ensure_rebalanced();
    // This function is idempotent
    co_await _hints_resource_manager.register_manager(_hints_manager);
    co_await _hints_manager.change_host_filter(std::move(new_filter));
}
```

From my POV it seems, that the logic related to directories initializer needs to be preserved.

pwrobelse replied:
I added a new function to `utils::directories` to encapsulate the creation of these three dirs.

xemul reviewed (changes requested):
The biggest essential concern is that database shouldn't act as a provider of directories for other services, they should get their reference on directories and use it

The technical merge-blocked is that this PR has all-in-one-patch and must be split into several smaller patches

xemul commented:
> Refs: #5626

Must be 'fixed', so that the issue gets automagically closed once this PR is merged

avikivity commented:
> This change:
> 
> * extends the public interface of utils::directories class to provide required directory paths to the users
> * removes 'db::config::setup_directories()' to avoid altering the fields of configuration object
> * replaces usages of db::config object with utils::directories object in places that require obtaining paths to dirs
> 
> Refs: #5626

Please explain why the whole thing is good (why, not just what).

pwrobelse commented:
> First things first -- please, split the change into several patches, it's very hard to review one huge change that does everything in one go

Fixed. The change now consist of a few small commits. Each of the them is described in commit message.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5920/)
#### Build Details:
- Duration: 8 min 45 sec
- Builder: spider1.cloudius-systems.com

pwrobelse commented:
> > This change:
> > 
> > * extends the public interface of utils::directories class to provide required directory paths to the users
> > * removes 'db::config::setup_directories()' to avoid altering the fields of configuration object
> > * replaces usages of db::config object with utils::directories object in places that require obtaining paths to dirs
> > 
> > Refs: #5626
> 
> Please explain why the whole thing is good (why, not just what).

Hi, I updated the description of the PR to add an explanation for the reasons. Is it okay now?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5923/)
#### Build Details:
- Duration: 10 min
- Builder: spider3.cloudius-systems.com

pwrobelse commented:
> > #### 🔴 CI State: FAILURE
> > ❌ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5920/)
> > #### Build Details:
> > ```
> > * Duration: 8 min 45 sec
> > 
> > * Builder: spider1.cloudius-systems.com
> > ```
> 
> At glance, this error does not make any sense to me. Firstly, the local build of the code passed. Secondly, the headers and forward declarations in `utils/directories` have not been changed between the current and the previous version.
> 
> `utils/directories.cc:80:30: error: member access into incomplete type 'const db::config'`
> 
> In `utils/directories.cc` I can see: `#include "db/config.hh"`
> 
> I will try to rerun the job.

It looks like a compiler issue. When I replaced the order of includes in `directories.cc` to include `db/config.hh` before `init.hh`, then `utils/directories.o` built without any problems.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5924/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7263/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5924/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
> > > This change:
> > > 
> > > * extends the public interface of utils::directories class to provide required directory paths to the users
> > > * removes 'db::config::setup_directories()' to avoid altering the fields of configuration object
> > > * replaces usages of db::config object with utils::directories object in places that require obtaining paths to dirs
> > > 
> > > Refs: #5626
> > 
> > 
> > Please explain why the whole thing is good (why, not just what).
> 
> Hi, I updated the description of the PR to add an explanation for the reasons. Is it okay now?

Yes, thanks.

bhalevy commented:
@xemul please re-review

denesb commented on tools/scylla-sstable.cc:
```diff
@@ -192,7 +193,9 @@ struct schema_with_source {
-std::optional<schema_with_source> try_load_schema_from_user_provided_source(const bpo::variables_map& app_config, db::config& cfg) {
+std::optional<schema_with_source> try_load_schema_from_user_provided_source(const bpo::variables_map& app_config,
+                                                                            db::config& cfg,
+                                                                            const ::utils::directories& dirs) {
```
nit: this kind of line-breaking is discouraged, as soon as the name of the method changes, the lines become unaligned.

Just align all wrapped lines, with two extra ident, compared to the original line.

pwrobelse replied:
Sure, fixed. Thanks :+1:

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6017/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6017/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7345/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider3.cloudius-systems.com

xemul commented on utils/directories.cc:
```diff
@@ -151,6 +151,17 @@ future<> directories::verify_owner_and_mode(fs::path path, recursive recursive)
+seastar::future<> directories::create_and_verify_directories() {
+    // Note: creation of hints_dir and view_hints_dir is
+    //       responsibility of db::hints::directory_initializer.
+    utils::directories::set dir_set;
```
Does the directories::set become internal class for utils/directories.cc after this change?

pwrobelse replied:
No, it does not. It is still used by:
`db::hints::directory_initializer::impl::ensure_created_and_verified()`

xemul replied:
OK, but it now looks like

```
        utils::directories::set dir_set;
        dir_set.add_sharded(_hints_directory);

        manager_logger.debug("Creating and validating hint directories: {}", _hints_directory);
        co_await _dirs.create_and_verify(std::move(dir_set));
```

just a single path. Worth encapsulating this class out of global scope

pwrobelse replied:
Okay, fixed. I made `utils::directories::set` an internal type and added a public method called `create_and_verify_sharded_directory()` that is used by `directory_initializer`.

xemul commented on utils/directories.cc:
```diff
@@ -77,9 +77,11 @@ void directories::set::add_sharded(sstring p) {
+directories::directories(const ::db::config& cfg)
+        : _developer_mode(cfg.developer_mode())
+{
+    setup_paths(cfg);
```
I don't think this call is needed. The setup_paths(cfg) does a bunch of `set_foo_dir(cfg.foo_directory())` calls and each `set_foo_dir()`, in turn, just initializes the corresponding `_foo_dir` member. It's better to initialize `_foo_dir`-s in constructor list-initializer directly

pwrobelse replied:
Moved to the initializer list and removed setters entirely.

xemul commented on utils/directories.cc:
```diff
@@ -190,6 +190,10 @@ static seastar::sstring to_sstring(const fs::path& p) {
     return seastar::sstring{p.c_str()};
 }
 
+seastar::sstring directories::get_work_dir() const {
```
This looks like runaway part from patch [Allow utils::directories to provide paths to dirs](https://github.com/scylladb/scylladb/pull/16787/commits/0b88fe8d6c7725899736b8900915ada7e1d7711a) of this PR. Please, either introduce _each_ `get_foo_dir()` getter one-by-one in the same patches that start using it, or introduce _all_ `get_foo_dir()` getters in one preparation patch

pwrobelse replied:
Sure, fixed. The content of this commit has been moved to the change, that adds the new interface. Thanks :+1:

xemul commented on utils/directories.cc:
```diff
@@ -149,4 +151,108 @@ future<> directories::verify_owner_and_mode(fs::path path, recursive recursive)
+    return to_sstring(_saved_caches_dir);
+}
+
+std::vector<seastar::sstring> directories::get_data_file_dirs() const {
```
I think that it's enough if this method returns const reference directly on _data_file_dirs. If caller is OK with using const reference, it will avoid copying of this vector, if not, it will just default-copy-construct it

pwrobelse replied:
Right now the paths are stored as member variables (`fs::path` type). Should we change them to `sstring`?

The issue contains the information that storing paths as `fs::path` should be considered. That's why I did it this way. But maybe sstring is sufficient.

Comment:
https://github.com/scylladb/scylladb/issues/5626#issuecomment-904695991

xemul commented on utils/directories.cc:
```diff
@@ -149,4 +151,108 @@ future<> directories::verify_owner_and_mode(fs::path path, recursive recursive)
+    _data_file_dirs = std::move(dirs_paths);
+}
+
+static seastar::sstring to_sstring(const fs::path& p) {
```
You have `using namespace seastar` up above, so the namespace specifier can be omitted here and all the way down

pwrobelse replied:
Fixed - thanks :+1:

xemul commented on replica/database.hh:
```diff
@@ -1749,6 +1752,7 @@ public:
     const db::config& get_config() const {
         return _cfg;
     }
+
```
nit: Try to avoid such hunks in the functional patches

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6077/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7400/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6077/)
#### Build Details:
- Duration: 2 hr 0 min
- Builder: spider1.cloudius-systems.com

xemul commented on main.cc:
```diff
@@ -967,13 +967,13 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
-            dir_set.add(cfg->schema_commitlog_directory());
+            dir_set.add(dirs->get_data_file_dirs());
+            dir_set.add(dirs->get_commitlog_dir());
+            dir_set.add(dirs->get_schema_commitlog_dir());
```
You remove this place in the next patch, so what's the point?

pwrobelse replied:
Fixed - this patch has been merged with the next one. Thanks!

xemul commented on utils/directories.cc:
```diff
@@ -131,7 +141,7 @@ void directories::override_if_empty(std::vector<fs::path>& v, const fs::path& de
 }
 
-future<> directories::create_and_verify(directories::set dir_set) {
+future<> directories::create_and_verify(directories::set&& dir_set) {
```
Why is this change in this patch? If it's copy optimization it can be done in its own patch

pwrobelse replied:
Fixed. I used a reference without much thinking - directories::set is only forward declared in the header file.

However, it is a private member function and it is used only in the source file. Without the reference it compiles. I reverted that change.

xemul commented:
General note -- no need in adding `refs` tag to every patch in the set, PR comment is enough

xemul commented:
Also, you mark #5626 with `refs` not `fixes`. What's left to be fixed to close the original issue?

pwrobelse commented:
> Also, you mark #5626 with `refs` not `fixes`. What's left to be fixed to close the original issue?

Replaced with `fixes`.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6140/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6140/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7440/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6180/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6180/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7482/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6180/)
#### Build Details:
- Duration: 4 hr 58 min
- Builder: i-037f333be9a6e6c04 (m5ad.8xlarge)

xemul commented:
I had to revert this PR due to #17078
